pkg/apis/kubeflow.org/v2alpha1/trainjob_types.go

```diff
@@ -15,3 +15,235 @@ limitations under the License.
+type Trainer struct {
+	// Docker image for the training container.
+	Image string `json:"image,omitempty"`
+
+	// Entrypoint commands for the training container.
+	Command []string `json:"command,omitempty"`
+
+	// Arguments to the entrypoint for the training container.
+	Args []string `json:"args,omitempty"`
+
+	// List of environment variables to set in the training container.
+	// These values will be merged with the TrainingRuntime's trainer environments.
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Number of training nodes.
+	// TODO (andreyvelich): Do we want to support dynamic num of nodes in TrainJob for PyTorch elastic: `--nnodes=1:4` ?
+	NumNodes *int32 `json:"numNodes,omitempty"`
+
+	// Compute resources for each training node.
+	ResourcesPerNode corev1.ResourceRequirements `json:"resourcesPerNode,omitempty"`
+
+	// Number of processes/workers/slots on every training node.
+	// For the Torch runtime: `auto`, `cpu`, `gpu`, or int value can be set.
+	// For the MPI runtime only int value can be set.
+	NumProcPerNode *string `json:"numProcPerNode,omitempty"`
+}
+
+// DatasetConfig represents the desired dataset configuration.
+// When this API is used, the training runtime must have
+// the `dataset-initializer` container in the `Initializer` Job.
+type DatasetConfig struct {
+	// Storage uri for the dataset provider.
+	StorageUri string `json:"storageUri"`
+
+	// List of environment variables to set in the dataset initializer container.
+	// These values will be merged with the TrainingRuntime's dataset initializer environments.
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Reference to the TrainJob's secrets to download dataset.
+	SecretRef corev1.SecretReference `json:"secretRef,omitempty"`
+}
+
+// ModelConfig represents the desired model configuration.
+type ModelConfig struct {
+	// Configuration of the pre-trained model.
+	// When this API is used, the training runtime must have
+	// the `model-initializer` container in the `Initializer` Job.
+	Input *InputModel `json:"input,omitempty"`
+
+	// Configuration of the trained model.
+	// When this API is used, the training runtime must have
+	// the `model-exporter` container in the `Exporter` Job.
+	Output *OutputModel `json:"output,omitempty"`
+}
+
+// InputModel represents the desired pre-trained model configuration.
+type InputModel struct {
+	// Storage uri for the model provider.
+	StorageUri string `json:"storageUri"`
+
+	// List of environment variables to set in the model initializer container.
+	// These values will be merged with the TrainingRuntime's model initializer environments.
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// Reference to the TrainJob's secrets to download model.
+	SecretRef corev1.SecretReference `json:"secretRef,omitempty"`
+}
+
+// OutputModel represents the desired trained model configuration.
+type OutputModel struct {
+	// Storage uri for the model exporter.
+	StorageUri string `json:"storageUri"`
```

Comment:
If we want to make this a required field, perhaps this is the right way to go - https://github.com/kubernetes-sigs/controller-tools/pull/944. (With `omitempty` for skipping serialization when unnecessary)

Reply:
I think, we should introduce +required and +optional parameters for kubebuilder as part of this issue: https://github.com/kubeflow/training-operator/issues/2219.
WDYT @tenzen-y @shravan-achar ?

Reply:
SGTM